app/concerns/long_runnable.rb

```diff
@@ -0,0 +1,64 @@
+module LongRunnable
+  extend ActiveSupport::Concern
+
+  included do |base|
+    AgentRunner.register(base)
+  end
+
+  module ClassMethods
+    def setup_worker
+      [new(id: self.to_s)]
+    end
+  end
+
+  class Worker
+    attr_reader :thread, :id, :agent, :config, :mutex
+
+    def initialize(options)
+      @id = options[:id]
+      @agent = options[:agent]
+      @config = options[:config]
+    end
+
+    def run!
+      @thread = Thread.new do
+        begin
+          run
+        rescue SignalException, SystemExit
+          stop!
+        rescue StandardError => e
+          STDERR.puts "\nException #{e.message}:\n#{e.backtrace.join("\n")}\n\n"
```

Comment:
Sounds like a good idea. I think we could work on the logging in general. The background worker have a lot of more or less random `puts` calls.